lib/matplotlib/collections.py

```diff
@@ -196,13 +219,30 @@ def get_datalim(self, transData):
             # get_path_collection_extents handles nan but not masked arrays
 
         if len(paths) and len(offsets):
-            result = mpath.get_path_collection_extents(
-                transform.frozen(), paths, self.get_transforms(),
-                offsets, transOffset.frozen())
-            result = result.inverse_transformed(transData)
-        else:
-            result = transforms.Bbox.null()
-        return result
+            if transform.contains_branch(transData):
+                # collections that are just in data units (like quiver)
+                # can properly have the axes limits set by their shape +
+                # offset.  LineCollections that have no offsets can
+                # also use this algorithm (like streamplot).
+                result = mpath.get_path_collection_extents(
+                    transform.frozen(), paths, self.get_transforms(),
+                    offsets, transOffset.frozen())
+                return result.inverse_transformed(transData)
+            if not self._offsetsNone:
+                # this is for collections that have their paths (shapes)
+                # in physical, axes-relative, or figure-relative units
+                # (i.e. like scatter). We can't uniquely set limits based on
+                # those shapes, so we just set the limits based on their
+                # location.
+                # Finish the transform:
```

Comment:
in the same spirit as https://github.com/matplotlib/matplotlib/pull/14845 this could be
```
offset_to_data = transOffset - transData
offsets = offset_to_data.transform(offsets)
```
(you also save a transform by doing so).

Reply:
I wasn't paying attention when #14845 went in.  This requires remembering that transA - transB is the same as what I wrote here.  I don't think it helps the readability/comprehensibility of all the transform magic to have arithmetic operators that do stuff versus just using explicit methods.

Reply:
The main operator is trA+trB, which just means "perform trA, then perform trB"; once you know it it's quite natural that trA-trB means "perform trA, then perform the inverse of trB".  You could write `mtransforms.composite_transform_factory(trA, trB.inverse())` if you prefer explicit method names, but the point is that it is more efficient to first multiply two small (3x3) matrices and then use the result to transform a big (N,2) array, rather than to do two transformations in a series.

Reply:
Good point.  Split the difference:  `(trA + trB.inverted())` seems straightforward enough.  Like, I understand the `__sub__` convention, I just find it a bit obscure.  The `__add__` convention is fine.